.github/workflows/update-plugins-repo-refs.yaml

```diff
@@ -63,243 +68,293 @@ jobs:
-          plugins=$(
-            comma=""
-            echo '['
-            for regexp in ${regexps}
+          minimalBackstageVersion=$(IFS=$'\n' && echo "${overlayRepoBranchToBackstageVersion[*]}" | sort -V | head -n1)
+          message "Minimal considered backstage version: ${minimalBackstageVersion}"
+
+          if [[ "${INPUT_SINGLE_BRANCH}" != "" ]]
+          then
+            singleBranchBackstageVersion="${overlayRepoBranchToBackstageVersion[${INPUT_SINGLE_BRANCH}]}"
+            overlayRepoBranchToBackstageVersion=()
+            overlayRepoBranchToBackstageVersion[${INPUT_SINGLE_BRANCH}]="${singleBranchBackstageVersion}"
+          fi
+          
+          message "Watched overlay branches:"
+          for overlayRepoBranch in "${!overlayRepoBranchToBackstageVersion[@]}"
+          do
+            message "  - ${overlayRepoBranch} -> Backstage ${overlayRepoBranchToBackstageVersion[${overlayRepoBranch}]}"
+          done
+
+          npmPackages=()
+          for regexp in ${INPUT_REGEXPS}
+          do
+            if [[ "${regexp}" == \'*\' ]]
+            then
+              message "Using raw plugin name: ${regexp}"
+              npmPackages+=("$(echo ${regexp} | sed -e 's/^.//' -e 's/.$//')")
+            else
+              message "Searching plugins for regexp: ${regexp}"
+              npmPackages+=($(npm search "/^(${regexp})/" --searchlimit=1000 --json --no-description --searchexclude "/^(?!(${regexp}))/" | jq -r '.[].name' | sort))
+            fi
+          done
+
+          missingPackages=()
+          publishedPlugins=()
+          processPackage() {
+            local packageName=$1
+            local searchForMissingPackages=$2
+            if [[ "${packageName}" == *"-node" ]] || [[ "${packageName}" == *"-common" ]] || [[ "${packageName}" == *"-react" ]] 
+            then
+              verbose "  Skipping published package ${packageName}: not a plugin"
+              return
+            fi
+            npmPackage=$(npm view --json ${packageName} name backstage versions)
+            if [[ "$(echo ${npmPackage} | jq -r '.backstage.role')" != *"-plugin"* ]]
+            then
+              verbose "  Skipping published package ${packageName}: not a plugin"
+              return
+            fi
+            firstVersionAnalyzed=false
+            message "  Fetching published versions of plugin ${packageName}"
+            for version in $(echo ${npmPackage} | jq -r '.versions | if type == "string" then . else .[] end | select(test("[^-]+-.+")|not)' | sort -Vr)
             do
-              if [[ "${regexp}" == \'*\' ]]
+              pluginInfo=$(npm view --json ${packageName}@${version} name version gitHead repository  | jq '. | {name, version, directory: (.repository.directory // "."), url: (.repository.url // ""), gitHead }')
+              gitRepo=$(echo ${pluginInfo} | jq -r '.url | 
+                if (. | startswith("https://github.com/"))
```

Comment:
I liked your other version of this with indents so it's easier to see the nested conditions